In [3]:
#reading multiple csv files
import pandas as pd
import glob #glob is amazing

csv_files = glob.glob(r"C:\Users\Abdul\Documents\VS_code_projects\OLIST_Data_Cleaning\data\raw\*.csv") #reads any file in the directory that ends with .csv

datasets = {}

for file in csv_files: 
    data = pd.read_csv(file) #read each file one by one
    datasets[file] = data #put each file into the dictionary so we can call those files without having to read each one again



In [4]:
#inspecting customers dataset

customers = datasets[r"C:\Users\Abdul\Documents\VS_code_projects\OLIST_Data_Cleaning\data\raw\olist_customers_dataset.csv"]

customers.head(10)

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP
5,879864dab9bc3047522c92c82e1212b8,4c93744516667ad3b8f1fb645a3116a4,89254,jaragua do sul,SC
6,fd826e7cf63160e536e0908c76c3f441,addec96d2e059c80c30fe6871d30d177,4534,sao paulo,SP
7,5e274e7a0c3809e14aba7ad5aae0d407,57b2a98a409812fe9618067b6b8ebe4f,35182,timoteo,MG
8,5adf08e34b2e993982a47070956c5c65,1175e95fb47ddff9de6b2b06188f7e0d,81560,curitiba,PR
9,4b7139f34592b3a31687243a302fa75b,9afe194fb833f79e300e37e580171f22,30575,belo horizonte,MG


**Customer id, customer_unique_id** look like an idetifier. not human readable.

Questions: 
- "is it unique?"
- "does it appear in other tables?"
- "does it have a consistent format?"
- "are there missing values?"
- "are there duplicates?"
- "why would a table need 2 ids?:

**Customer_zip_code_prefix** - in the data head itself i notice two values 1151 and 4534 having fewer than 5 digits. 

lets inpect another table 

In [5]:
#inspectin orders_dataset

orders = datasets[r"C:\Users\Abdul\Documents\VS_code_projects\OLIST_Data_Cleaning\data\raw\olist_orders_dataset.csv"]

orders.head(10)

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00
5,a4591c265e18cb1dcee52889e2d8acc3,503740e9ca751ccdda7ba28e9ab8f608,delivered,2017-07-09 21:57:05,2017-07-09 22:10:13,2017-07-11 14:58:04,2017-07-26 10:57:55,2017-08-01 00:00:00
6,136cce7faa42fdb2cefd53fdc79a6098,ed0271e0b7da060a393796590e7b737a,invoiced,2017-04-11 12:22:08,2017-04-13 13:25:17,NaN,NaN,2017-05-09 00:00:00
7,6514b8ad8028c9f2cc2374ded245783f,9bdf08b4b3b52b5526ff42d37d47f222,delivered,2017-05-16 13:10:30,2017-05-16 13:22:11,2017-05-22 10:07:46,2017-05-26 12:55:51,2017-06-07 00:00:00
8,76c6e866289321a7c93b82b54852dc33,f54a9f0e6b351c431402b8461ea51999,delivered,2017-01-23 18:29:09,2017-01-25 02:50:47,2017-01-26 14:16:31,2017-02-02 14:08:10,2017-03-06 00:00:00
9,e69bfb5eb88e0ed6a785585b27e16dbf,31ad1d1b63eb9962463f764d4e6e0c9d,delivered,2017-07-29 11:55:02,2017-07-29 12:05:32,2017-08-10 19:45:24,2017-08-16 17:14:30,2017-08-23 00:00:00


**order_status** is important, it describes the state of an order
the date columns need to be examined for consistency and datatype
NaN values need to be quantified, not just noticed visually
the different date columns probably have difference business meaning so we shouldnt assume they all behave identically

In [6]:
orders.info()
orders.isnull().sum()

<class 'pandas.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype
---  ------                         --------------  -----
 0   order_id                       99441 non-null  str  
 1   customer_id                    99441 non-null  str  
 2   order_status                   99441 non-null  str  
 3   order_purchase_timestamp       99441 non-null  str  
 4   order_approved_at              99281 non-null  str  
 5   order_delivered_carrier_date   97658 non-null  str  
 6   order_delivered_customer_date  96476 non-null  str  
 7   order_estimated_delivery_date  99441 non-null  str  
dtypes: str(8)
memory usage: 6.1 MB


order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

**IN ORDERS DATASET**      

- `customer_id` appears in both the customers and orders tables.
- `order_id` is present in the orders table.
- Several timestamp/date columns are present.
- Some date columns contain missing values.
- All columns are currently stored as strings.


**MISSING VALUES:**

`order_approved_at` = 99441 - 99281 = 160

`order_delivered_carrirer_date` 99441 - 97658 =  1783

`order_delivered_customer_date` 99441 - 96476 = 2965


**Questions:**

- "How do we fill these NaNs?"
- "Why are these values missing?"
- "Are missing delivery dates associated with "order status" values?"
- "Are all timestamp columns consistently formatted?"

if an order hasnt been delivered, it might naturally have no value.

In [7]:
orders["order_status"].value_counts()

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

### Missing delivery date investigation

- `order_status` contains 96,478 `delivered` orders out of 99,441 total orders.
- `order_delivered_customer_date` has 2,965 missing values.
- There are 2,963 non-delivered orders:
  `99,441 - 96,478 = 2,963`
- Therefore, 2 missing `order_delivered_customer_date` values appear to belong to orders marked as `delivered`.

### Question

- Why do these 2 orders have `order_status = delivered` but no `order_delivered_customer_date`?
- Need to inspect the 2 records before deciding whether they are data-quality errors.

In [8]:
orders[orders["order_delivered_customer_date"].isnull() & (orders["order_status"] == "delivered")]

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
3002,2d1e2d5bf4dc7227b3bfebb81328c15f,ec05a6d8558c6455f0cbbd8a420ad34f,delivered,2017-11-28 17:44:07,2017-11-28 17:56:40,2017-11-30 18:12:23,NaN,2017-12-18 00:00:00
20618,f5dd62b788049ad9fc0526e3ad11a097,5e89028e024b381dc84a13a3570decb4,delivered,2018-06-20 06:58:43,2018-06-20 07:19:05,2018-06-25 08:05:00,NaN,2018-07-16 00:00:00
43834,2ebdfc4f15f23b91474edf87475f108e,29f0540231702fda0cfdee0a310f11aa,delivered,2018-07-01 17:05:11,2018-07-01 17:15:12,2018-07-03 13:57:00,NaN,2018-07-30 00:00:00
79263,e69f75a717d64fc5ecdfae42b2e8e086,cfda40ca8dd0a5d486a9635b611b398a,delivered,2018-07-01 22:05:55,2018-07-01 22:15:14,2018-07-03 13:57:00,NaN,2018-07-30 00:00:00
82868,0d3268bad9b086af767785e3f0fc0133,4f1d63d35fb7c8999853b2699f5c7649,delivered,2018-07-01 21:14:02,2018-07-01 21:29:54,2018-07-03 09:28:00,NaN,2018-07-24 00:00:00
92643,2d858f451373b04fb5c984a1cc2defaf,e08caf668d499a6d643dafd7c5cc498a,delivered,2017-05-25 23:22:43,2017-05-25 23:30:16,NaN,NaN,2017-06-23 00:00:00
97647,ab7c89dc1bf4a1ead9d6ec1ec8968a84,dd1b84a7286eb4524d52af4256c0ba24,delivered,2018-06-08 12:09:39,2018-06-08 12:36:39,2018-06-12 14:10:00,NaN,2018-06-26 00:00:00
98038,20edc82cf5400ce95e1afacc25798b31,28c37425f1127d887d7337f284080a0f,delivered,2018-06-27 16:09:12,2018-06-27 16:29:30,2018-07-03 19:26:00,NaN,2018-07-19 00:00:00


In [9]:
pd.crosstab(
    orders["order_status"],
    orders["order_delivered_customer_date"].isnull()
)

order_delivered_customer_date,False,True
order_status,,
approved,0,2
canceled,6,619
created,0,5
delivered,96470,8
invoiced,0,314
processing,0,301
shipped,0,1107
unavailable,0,609


### Missing delivery date investigation

- `order_delivered_customer_date` has 2,965 missing values.
- Filtering for orders where:
  - `order_status == "delivered"`
  - `order_delivered_customer_date` is missing
- returned 8 records.
- Therefore, missing delivery dates are not exclusively associated with non-delivered orders.

### Question

- Why are these 8 orders marked as `delivered` despite having no customer delivery date?
- Need to investigate these records and determine whether they represent legitimate missing data or data-quality issues.

In [10]:
orders["order_purchase_timestamp"].tail(20)


99421    2017-07-05 15:55:19
99422    2018-05-25 17:05:17
99423    2017-10-12 20:54:11
99424    2017-11-21 20:18:03
99425    2018-03-21 19:47:18
99426    2017-05-15 09:46:26
99427    2018-05-12 13:42:24
99428    2018-06-22 20:53:29
99429    2017-11-05 16:17:53
99430    2018-03-04 22:48:38
99431    2017-10-27 15:21:00
99432    2017-12-20 09:52:41
99433    2017-10-04 19:57:37
99434    2017-01-27 00:30:03
99435    2017-02-23 09:05:12
99436    2017-03-09 09:54:05
99437    2018-02-06 12:58:58
99438    2017-08-27 14:46:43
99439    2018-01-08 21:28:27
99440    2018-03-08 20:57:30
Name: order_purchase_timestamp, dtype: str

In [11]:
pd.to_datetime(orders["order_purchase_timestamp"], errors = 'coerce')
orders["order_purchase_timestamp"].isnull().sum()

np.int64(0)

### `order_purchase_timestamp`

- Stored as `str` in the raw dataset.
- Values follow the apparent format `YYYY-MM-DD HH:MM:SS`.
- `pd.to_datetime(..., errors="coerce")` successfully parsed all 99,441 values.
- 0 values became `NaT`.
- No invalid/unparseable values detected in this column.

In [12]:
orders.columns

Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date'],
      dtype='str')

In [13]:
date_columns = ['order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date']

for column in date_columns:
    converted = pd.to_datetime(orders[column], errors = "coerce")
    print(column, converted.isnull().sum())



order_purchase_timestamp 0
order_approved_at 160
order_delivered_carrier_date 1783
order_delivered_customer_date 2965
order_estimated_delivery_date 0


### Date columns

All date columns are stored as `str` in the raw dataset.

After testing conversion with `pd.to_datetime(..., errors="coerce")`:

| Column | Failed/NaT values |
|---|---|
| `order_purchase_timestamp` | 0 |
| `order_approved_at` | 160 |
| `order_delivered_carrier_date` | 1,783 |
| `order_delivered_customer_date` | 2,965 |
| `order_estimated_delivery_date` | 0 |

Observations:
- No invalid date strings were detected.
- The missing values in the three affected columns remain missing after conversion.
- Date values appear to follow a consistent `YYYY-MM-DD HH:MM:SS` format.
- These columns can likely be converted to datetime during the cleaning phase.

In [14]:
orders.order_id.nunique()

99441

### `order_id`

- Total rows: 99,441
- Unique `order_id` values: 99,441
- Every row has a unique `order_id`.
- `order_id` appears to function as the unique identifier for each order.

In [15]:
orders["customer_id"].isin(customers["customer_id"]).value_counts()

customer_id
True    99441
Name: count, dtype: int64

### Referential Integrity

Checks whether every `customer_id` in `orders` exists in `customers`.

```python
orders["customer_id"].isin(customers["customer_id"]).value_counts()

safe to assume customer.customer_id = orders.customer_id

In [16]:
customers["customer_unique_id"].value_counts().head(10)

customer_unique_id
8d50f5eadf50201ccdcedfb9e2ac8455    17
3e43e6105506432c953e165fb2acf44c     9
1b6c7548a2a1f9037c1fd3ddfed95f33     7
6469f99c1f9dfae7733b25662e7f1782     7
ca77025e7201e3b30c44b472ff346268     7
47c1a3033b8b77b3ab6e109eb4d5fdf3     6
dc813062e0fc23409cd255f7f53c7074     6
63cfc61cee11cbe306bff5857d00bfe4     6
f0e310a6839dce9de1638e0fe5ab282a     6
de34b16117594161a6a89c50b289d35a     6
Name: count, dtype: int64

In [17]:
customers[customers["customer_unique_id"] == "8d50f5eadf50201ccdcedfb9e2ac8455"]

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
14186,1bd3585471932167ab72a84955ebefea,8d50f5eadf50201ccdcedfb9e2ac8455,4045,sao paulo,SP
15321,a8fabc805e9a10a3c93ae5bff642b86b,8d50f5eadf50201ccdcedfb9e2ac8455,4045,sao paulo,SP
16654,897b7f72042714efaa64ac306ba0cafc,8d50f5eadf50201ccdcedfb9e2ac8455,4045,sao paulo,SP
36122,b2b13de0770e06de50080fea77c459e6,8d50f5eadf50201ccdcedfb9e2ac8455,4045,sao paulo,SP
38073,42dbc1ad9d560637c9c4c1533746f86d,8d50f5eadf50201ccdcedfb9e2ac8455,4045,sao paulo,SP
40141,dfb941d6f7b02f57a44c3b7c3fefb44b,8d50f5eadf50201ccdcedfb9e2ac8455,4045,sao paulo,SP
48614,65f9db9dd07a4e79b625effa4c868fcb,8d50f5eadf50201ccdcedfb9e2ac8455,4045,sao paulo,SP
52574,1c62b48fb34ee043310dcb233caabd2e,8d50f5eadf50201ccdcedfb9e2ac8455,4045,sao paulo,SP
58707,a682769c4bc10fc6ef2101337a6c83c9,8d50f5eadf50201ccdcedfb9e2ac8455,4045,sao paulo,SP
67996,6289b75219d757a56c0cce8d9e427900,8d50f5eadf50201ccdcedfb9e2ac8455,4045,sao paulo,SP


In [18]:
customer = customers [customers["customer_unique_id"] == customers["customer_unique_id"].value_counts().index[0]] 

customer

customer_orders = orders[orders["customer_id"].isin(customer["customer_id"])

]

customer_orders

len(customer)


17

In [19]:
customers.groupby("customer_unique_id")["customer_id"].nunique().value_counts().sort_index() #group all customer_ids to customer_unique_id, count their unique values, and sort them

customer_id
1     93099
2      2745
3       203
4        30
5         8
6         6
7         3
9         1
17        1
Name: count, dtype: int64

### Inspection Flow
customer_id
    ↓

Is it unique?
    ↓

Yes, 99,441 unique IDs
    ↓

customer_unique_id
    ↓

Is IT unique?
    ↓

No, 96,096 unique IDs
    ↓
    
So some customer_unique_id values repeat
    ↓ 

How much do they repeat?
    ↓

value_counts()
    ↓

One customer appears 17 times
    ↓

Is this just one weird customer?
    ↓

We need to examine ALL customers
    ↓
    
groupby()

In [20]:
orders.groupby("customer_id")["order_id"].nunique().value_counts().sort_index()

order_id
1    99441
Name: count, dtype: int64

### Customer ID relationship

- Every `customer_id` in `orders` is associated with exactly one `order_id`.
- `customer_unique_id` can appear across multiple `customer_id` records.
- This indicates that multiple orders can belong to the same underlying customer.
- No cleaning is required for these repeated `customer_unique_id` values.

In [21]:
customers.groupby("customer_zip_code_prefix")["customer_state"].nunique().value_counts().sort_index()

customer_state
1    14994
Name: count, dtype: int64

### Geographic consistency

- Checked the relationship between `customer_zip_code_prefix` and `customer_state`.
- All 14,994 ZIP-code prefixes are associated with exactly one state.
- No ZIP/state conflicts were found.

In [22]:
customers.groupby("customer_zip_code_prefix")["customer_city"].nunique().value_counts().sort_index()

customer_city
1    14955
2       38
3        1
Name: count, dtype: int64

In [23]:
customers.groupby("customer_zip_code_prefix")["customer_state"].nunique().value_counts()

customer_state
1    14994
Name: count, dtype: int64

In [24]:
customers.groupby("customer_zip_code_prefix")["customer_city"].nunique().value_counts()

customer_city
1    14955
2       38
3        1
Name: count, dtype: int64

In [25]:
conflicting_zips = (customers.groupby("customer_zip_code_prefix")["customer_city"].nunique())
conflicting_zips = conflicting_zips[conflicting_zips>1].index

customers[customers["customer_zip_code_prefix"].isin(conflicting_zips)][["customer_zip_code_prefix", "customer_city", 'customer_state']]

,customer_zip_code_prefix,customer_city,customer_state
54,23970,parati,RJ
60,72270,brasilia,DF
142,6835,embu,SP
633,6824,embu das artes,SP
695,85139,vitoria,PR
...,...,...,...
98831,15720,palmeira d'oeste,SP
98975,6810,embu das artes,SP
99135,6806,embu,SP
99367,73060,brasilia,DF


## Conclusion

The initial raw data inspection is complete for the `customers` and `orders` datasets.

Key findings:
- Both datasets contain the expected columns and records.
- No missing values were found in `customers`.
- Missing values were found in several `orders` date columns.
- All order purchase timestamps were successfully parsed as dates.
- `customer_id` values are consistent between `customers` and `orders`.
- Each `customer_id` is associated with one order in this dataset.
- `customer_unique_id` reveals that some customers placed multiple orders.
- Some ZIP codes contain multiple city names, which may represent naming inconsistencies and will be considered during cleaning.

The remaining Olist datasets still need to be inspected before the cleaning phase begins.

Next step: continue the raw data inspection with `order_items`.

In [26]:
order_items = datasets[r"C:\Users\Abdul\Documents\VS_code_projects\OLIST_Data_Cleaning\data\raw\olist_order_items_dataset.csv"]
order_items


,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14
...,...,...,...,...,...,...,...
112645,fffc94f6ce00a00581880bf54a75a037,1,4aa6014eceb682077f9dc4bffebc05b0,b8bc237ba3788b23da09c0f1f3a3288c,2018-05-02 04:11:01,299.99,43.41
112646,fffcd46ef2263f404302a634eb57f7eb,1,32e07fd915822b0765e448c4dd74c828,f3c38ab652836d21de61fb8314b69182,2018-07-20 04:31:48,350.00,36.53
112647,fffce4705a9662cd70adb13d4a31832d,1,72a30483855e2eafc67aee5dc2560482,c3cfdc648177fdbbbb35635a37472c53,2017-10-30 17:14:25,99.90,16.95
112648,fffe18544ffabc95dfada21779c9644f,1,9c422a519119dcad7575db5af1ba540e,2b3e4a2a3ea8e01938cabda2a3e5cc79,2017-08-21 00:04:32,55.99,8.72


### Columns and Data Types

- 7 columns were found.
- `shipping_limit_date` is stored as a string and will be considered for date conversion during cleaning.
- `price` and `freight_value` are stored as `float64`.
- `order_item_id` is stored as `int64`.
- ID columns are stored as strings.

In [27]:
order_items


,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14
...,...,...,...,...,...,...,...
112645,fffc94f6ce00a00581880bf54a75a037,1,4aa6014eceb682077f9dc4bffebc05b0,b8bc237ba3788b23da09c0f1f3a3288c,2018-05-02 04:11:01,299.99,43.41
112646,fffcd46ef2263f404302a634eb57f7eb,1,32e07fd915822b0765e448c4dd74c828,f3c38ab652836d21de61fb8314b69182,2018-07-20 04:31:48,350.00,36.53
112647,fffce4705a9662cd70adb13d4a31832d,1,72a30483855e2eafc67aee5dc2560482,c3cfdc648177fdbbbb35635a37472c53,2017-10-30 17:14:25,99.90,16.95
112648,fffe18544ffabc95dfada21779c9644f,1,9c422a519119dcad7575db5af1ba540e,2b3e4a2a3ea8e01938cabda2a3e5cc79,2017-08-21 00:04:32,55.99,8.72


### Observations 
112650 entries


In [66]:
order_items["order_id"].value_counts()

order_id
8272b63d03f5f79c56e9e4120aec44ef    21
1b15974a0141d54e36626dca3fdc731a    20
ab14fdcfbe524636d65ee38360e22ce8    20
428a2f660dc84138d969ccd69a0ab6d5    15
9ef13efd6949e4573a18964dd1bbe7f5    15
                                    ..
fffc94f6ce00a00581880bf54a75a037     1
fffcd46ef2263f404302a634eb57f7eb     1
fffce4705a9662cd70adb13d4a31832d     1
fffe18544ffabc95dfada21779c9644f     1
fffe41c64501cc87c801fd61db3f6244     1
Name: count, Length: 98666, dtype: int64

In [70]:
order_items[order_items["order_id"] == "8272b63d03f5f79c56e9e4120aec44ef"][["order_id", "order_item_id"]]

,order_id,order_item_id
57297,8272b63d03f5f79c56e9e4120aec44ef,1
57298,8272b63d03f5f79c56e9e4120aec44ef,2
57299,8272b63d03f5f79c56e9e4120aec44ef,3
57300,8272b63d03f5f79c56e9e4120aec44ef,4
57301,8272b63d03f5f79c56e9e4120aec44ef,5
57302,8272b63d03f5f79c56e9e4120aec44ef,6
57303,8272b63d03f5f79c56e9e4120aec44ef,7
57304,8272b63d03f5f79c56e9e4120aec44ef,8
57305,8272b63d03f5f79c56e9e4120aec44ef,9
57306,8272b63d03f5f79c56e9e4120aec44ef,10


### Observation

`order_item_id` shows incremental behavior within each `order_id`.

For example, an order with multiple items has `order_item_id` values progressing from 1, 2, 3, and so on.

This suggests that `order_item_id` represents the item's position within an order rather than a globally unique identifier.

In [62]:
order_range = order_items.groupby("order_id")["order_item_id"].agg(["count", "max"])
order_range[order_range["count"] > 1]

,count,max
order_id,,
0008288aa423d2a3f00fcb17cd7d8719,2,2
00143d0f86d6fbd9f9b38ab440ac16f5,3,3
001ab0a7578dd66cd4b0a71f5b6e1e41,3,3
001d8f0e34a38c37f7dba2a37d4eba8b,2,2
002c9def9c9b951b1bec6d50753c9891,2,2
...,...,...
ffd84ab39cd5e873d8dba24342e65c01,2,2
ffe4b41e99d39f0b837a239110260530,2,2
ffecd5a79a0084f6a592288c67e3c298,3,3


In [58]:
order_range = order_items.groupby("order_id")["order_item_id"].agg(["count", "max"])

order_range[order_range["count"] != order_range["max"]].head(40)

,count,max
order_id,,


### Order Item Sequence

- `order_item_id` is sequential within each `order_id`.
- The combination of `order_id` and `order_item_id` contains no duplicates.
- For every order, the number of item rows matches the maximum `order_item_id`.
- No sequence inconsistencies were found.

In [71]:
order_items["order_id"].isin(orders["order_id"]).value_counts()

order_id
True    112650
Name: count, dtype: int64

In [82]:
products = datasets[r"C:\Users\Abdul\Documents\VS_code_projects\OLIST_Data_Cleaning\data\raw\olist_products_dataset.csv"]

order_items["product_id"].isin(products["product_id"]).value_counts()

product_id
True    112650
Name: count, dtype: int64

## Order Items

### Observations

- 112,650 rows and 7 columns.
- No missing values were found.
- `shipping_limit_date` is stored as a string and should be considered for date conversion during cleaning.
- `order_id` is not unique because one order can contain multiple items.
- `order_item_id` behaves sequentially within each `order_id`.
- No duplicate `order_id` + `order_item_id` combinations were found.
- All `order_id` values exist in the `orders` dataset.
- All `product_id` values exist in the `products` dataset.

In [88]:
order_payments = datasets[r"C:\Users\Abdul\Documents\VS_code_projects\OLIST_Data_Cleaning\data\raw\olist_order_payments_dataset.csv"]

order_payments.shape

(103886, 5)

In [89]:
order_payments.info()

<class 'pandas.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  str    
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  str    
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), str(2)
memory usage: 4.0 MB


In [93]:
order_payments["order_id"].isin(orders["order_id"]).value_counts()

order_id
True    103886
Name: count, dtype: int64

In [94]:
order_payments.head(10)

,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71
3,ba78997921bbcdc1373bb41e913ab953,1,credit_card,8,107.78
4,42fdf880ba16b47b59251dd489d4441a,1,credit_card,2,128.45
5,298fcdf1f73eb413e4d26d01b25bc1cd,1,credit_card,2,96.12
6,771ee386b001f06208a7419e4fc1bbd7,1,credit_card,1,81.16
7,3d7239c394a212faae122962df514ac7,1,credit_card,3,51.84
8,1f78449c87a54faf9e96e88ba1491fa9,1,credit_card,6,341.09
9,0573b5e23cbd798006520e1d5b4c6714,1,boleto,1,51.95


In [102]:
payment_sequence = order_payments.groupby("order_id")["payment_sequential"].agg(["count", "max"])

payment_sequence[(payment_sequence["count"] ==  payment_sequence["max"]) & (payment_sequence["count"] > 1)].head(19)

,count,max
order_id,,
0016dfedd97fc2950e388d2971d718c7,2,2
002f19a65a2ddd70a090297872e6d64e,2,2
0071ee2429bc1efdc43aa3e073a5290e,2,2
009ac365164f8e06f59d18a08045f6c4,6,6
00b4a910f64f24dbcac04fe54088a443,2,2
00bd50cdd31bd22e9081e6e2d5b3577b,3,3
00c405bd71187154a7846862f585a9d4,7,7
00c95282163553a982f38481f9488481,4,4
00e6bc6b166eb28b4502c1cad4457248,2,2


In [103]:
order_payments[["order_id", "payment_sequential"]].duplicated().sum()

np.int64(0)

In [105]:
order_payments["payment_type"].value_counts()

payment_type
credit_card    76795
boleto         19784
voucher         5775
debit_card      1529
not_defined        3
Name: count, dtype: int64

In [107]:
order_payments[order_payments["payment_type"] == "not_defined"]

,order_id,payment_sequential,payment_type,payment_installments,payment_value
51280,4637ca194b6387e2d538dc89b124b0ee,1,not_defined,1,0.0
57411,00b1cb0320190ca0daa2c88b35206009,1,not_defined,1,0.0
94427,c8c528189310eaa44a745b8d9d26908b,1,not_defined,1,0.0
